polonius-engine/src/output/mod.rs

```diff
@@ -81,54 +93,179 @@ pub struct Output<T: FactTypes> {
+    region_live_at: Relation<(T::Origin, T::Point)>,
+    invalidates: Relation<(T::Loan, T::Point)>,
+
+    // static inputs used via `Variable`s, by all variants
+    outlives: &'ctx Vec<(T::Origin, T::Origin, T::Point)>,
+    borrow_region: &'ctx Vec<(T::Origin, T::Loan, T::Point)>,
+
+    // static inputs used by variants other than `LocationInsensitive`
+    killed: Relation<(T::Loan, T::Point)>,
+
+    // while this static input is unused by `LocationInsensitive`, it's depended on by initialization
+    // and liveness, so already computed by the time we get to borrowcking.
+    cfg_edge: Relation<(T::Point, T::Point)>,
+
+    // Partial results possibly used by other variants as input
+    potential_errors: Option<FxHashSet<T::Loan>>,
 }
 
 impl<T: FactTypes> Output<T> {
+    /// All variants require the same initial preparations, done in multiple
+    /// successive steps:
+    /// - compute initialization data
+    /// - compute liveness
+    /// - prepare static inputs as shared `Relation`s
+    /// - in cases where `LocationInsensitive` variant is ran as a filtering pre-pass,
+    ///   partial results can also be stored in the context, so that the following
+    ///   variant can use it to prune its own input data
     pub fn compute(all_facts: &AllFacts<T>, algorithm: Algorithm, dump_enabled: bool) -> Self {
-        match algorithm {
-            Algorithm::Naive => naive::compute(dump_enabled, all_facts.clone()),
-            Algorithm::DatafrogOpt => datafrog_opt::compute(dump_enabled, all_facts.clone()),
-            Algorithm::LocationInsensitive => {
-                location_insensitive::compute(dump_enabled, &all_facts)
+        let mut result = Output::new(dump_enabled);
+
+        // TODO: remove all the cloning thereafter, but that needs to be done in concert with rustc
+
+        let cfg_edge = all_facts.cfg_edge.clone().into();
+
+        // 1) Initialization
+        let initialization_ctx = InitializationContext {
+            child: all_facts.child.clone(),
+            path_belongs_to_var: all_facts.path_belongs_to_var.clone(),
+            initialized_at: all_facts.initialized_at.clone(),
+            moved_out_at: all_facts.moved_out_at.clone(),
+            path_accessed_at: all_facts.path_accessed_at.clone(),
+        };
+
+        let var_maybe_initialized_on_exit = initialization::init_var_maybe_initialized_on_exit(
+            initialization_ctx,
+            &cfg_edge,
+            &mut result,
+        );
+
+        // 2) Liveness
+        let universal_regions = all_facts.universal_region.clone();
+
+        let liveness_ctx = LivenessContext {
+            var_used: all_facts.var_used.clone(),
+            var_defined: all_facts.var_defined.clone(),
+            var_drop_used: all_facts.var_drop_used.clone(),
+            var_uses_region: all_facts.var_uses_region.clone(),
+            var_drops_region: all_facts.var_drops_region.clone(),
+        };
```

Comment:
Why do you need these clones? I thought they were only used here?

Reply:
For the same reason as the previous `mem::replace`: we can't move these out of `all_facts: &AllFacts<T>`